File: high-gas-py/src/agent.py
from forta_agent import Finding, FindingType, FindingSeverity, get_transaction_receipt

MEDIUM_GAS_THRESHOLD = 1000000
HIGH_GAS_THRESHOLD = 3000000
CRITICAL_GAS_THRESHOLD = 7000000

findings_count = 0


def provide_handle_transaction(get_transaction_receipt):
    def handle_transaction(transaction_event):
        # limiting this agent to emit only 5 findings so that the alert feed is not spammed
        global findings_count
        if findings_count >= 5:
            return []

        findings = []

        receipt = get_transaction_receipt(transaction_event.hash)
        gas_used = int(receipt.gas_used)
        if gas_used < MEDIUM_GAS_THRESHOLD:
            return findings

        findings.append(Finding({
            'name': 'High Gas Used',
            'description': f'Gas Used: {gas_used}',
            'alert_id': 'FORTA-1',
            'type': FindingType.Suspicious,
            'severity': get_severity(gas_used),
            'metadata': {
                'gas_used': gas_used
            }
        }))
        findings_count += len(findings)
        return findings

    return handle_transaction


def get_severity(gas_used):
    if gas_used > CRITICAL_GAS_THRESHOLD:
        return FindingSeverity.Critical
    elif gas_used > HIGH_GAS_THRESHOLD:
        return FindingSeverity.High
    else:
        return FindingSeverity.Medium


real_handle_transaction = provide_handle_transaction(get_transaction_receipt)


def handle_transaction(transaction_event):
    return real_handle_transaction(transaction_event)

File: hiding-sensitive-data-js/src/agent.js
const BigNumber = require("bignumber.js");
const { Finding, FindingSeverity, FindingType, getTransactionReceipt } = require("forta-agent");

let findingsCount = 0;
// this is the sensitive data we want to obfuscate
const INFURA_API_KEY = "97d4fafcc17b46819295fa833ac8877d";

function provideHandleTransaction(getTransactionReceipt) {
  return async function handleTransaction(txEvent) {
    const findings = [];
  
    // limiting this agent to emit only 5 findings so that the alert feed is not spammed
    if (findingsCount >= 5) return findings;
  
    // create finding if gas used is higher than threshold
    const receipt = await getTransactionReceipt(txEvent.hash)
    const gasUsed = new BigNumber(receipt.gasUsed)
    if (gasUsed.isGreaterThan("1000000")) {
      findings.push(
        Finding.fromObject({
          name: "High Gas Used",
          description: `Gas Used: ${gasUsed}`,
          alertId: "FORTA-1",
          severity: FindingSeverity.Medium,
          type: FindingType.Suspicious,
        })
      );
      findingsCount++;
    }
  
    return findings;
  };
}

module.exports = {
  provideHandleTransaction,
  handleTransaction: provideHandleTransaction(getTransactionReceipt),
};
